Python/ceval_macros.h

```diff
@@ -245,14 +282,24 @@ GETITEM(PyObject *v, Py_ssize_t i) {
 #define UPDATE_MISS_STATS(INSTNAME) ((void)0)
 #endif
 
-#define DEOPT_IF(COND, INSTNAME)                            \
-    if ((COND)) {                                           \
-        /* This is only a single jump on release builds! */ \
-        UPDATE_MISS_STATS((INSTNAME));                      \
-        assert(_PyOpcode_Deopt[opcode] == (INSTNAME));      \
-        goto PREDICTED_##INSTNAME;                          \
-    }
-
+#ifdef Py_TAIL_CALL_INTERP
+#   define GO_TO_INSTRUCTION_IF(COND, INSTNAME, SIZE)           \
+        if ((COND)) {                                           \
+            /* This is only a single jump on release builds! */ \
+            UPDATE_MISS_STATS((INSTNAME));                      \
+            assert(_PyOpcode_Deopt[opcode] == (INSTNAME));      \
+            Py_MUSTTAIL                                         \
+            return (INSTRUCTION_TABLE[INSTNAME])(frame, stack_pointer, tstate, next_instr - 1 - SIZE, opcode, oparg); \
```

Comment:
```suggestion
            return _TAIL_CALL_ ## INSTNAME(frame, stack_pointer, tstate, this_instr, opcode, oparg); \
```
I don't think this needs to go via the instruction table. And you don't need `SIZE` if you use `this_instr`.
You'll need to change the code generator so that it knows the use of `GO_TO_INSTRUCTION_IF` implies `this_instr`.

Reply:
Thanks!